### ISSUE TITLE
[Question] Docker official images vs. distro official repos.

### ISSUE THREAD

**Issue by pnloyd** [docker-library/official-images#3517]
I see that many (or maybe all?) docker images are building the software they distribute from scratch, even when there are packages available in the base image distro repos. I'd understand some benefits in the cases of packages that come from larger distros to save space, but in the case of a "micro" distro like alpine it makes less sense to me, where I imagine that their packages have already been optimized size.

My questions stems from my desire to have an image that **contains both php and nginx** running on alpine. I can make this myself without much trouble by installing from distro repos, but this leaves me wondering what I am missing out on because I cannot easilly combine the custom built official nginx install and custom built official PHP install because they live in separate images.

**TLDR:** I think the essence of my question is, "what is the underlying goal/advantage of creating/using custom installs as apposed to using upstream packages?"

Sorry if this is the wrong place to be asking this, I couldn't find an explanation in the docs.

**Comment by pnloyd** [comment]
Closing because I'm starting to understand the answer somewhat..

**Comment by tianon** [comment]
There's not any clear answer on which to prefer simply because we defer to upstream's official recommendations on a case-by-case basis.  For example, in the `postgres` image's Debian variants, we consume the `.deb` packages built and published by the upstream project, but in the Alpine variants we have to build from source (since upstream does not publish Alpine packages and instead recommends building from source for platforms for which they do not publish their own packages).

In `httpd`, Apache upstream does not publish any binary packages, and we thus build from source.

In `openjdk`, we install packages built by Debian, Alpine, etc. simply because OpenJDK itself is very difficult and time-consuming to build from source, and because upstream officially recommends those packages as a suitable source of OpenJDK.

Hope that helps!

**Comment by pnloyd** [comment]
Thanks I think I understand, there's a kind of quality control going on here by deferring to upstream.

**Comment by tianon** [comment]
See also https://github.com/docker-library/faq#why-do-so-many-official-images-build-from-source! :tada:
